core/base/inc/TMathBase.h

```diff
@@ -336,7 +336,7 @@ Iterator TMath::BinarySearch(Iterator first, Iterator last, Element value)
    if ( (pind != last) && (*pind == value) )
       return pind;
    else
-      return ( pind - 1);
+      return ((pind == first) ? pind : pind - 1);
```

Comment:
This does not seem to match the contract. i.e. if the smallest/first element is larger than the value, it still returns 'first'.   Since `std::lower_bound` already return `last`, I think the right code might be:
```suggestion
      return ((pind == first) ? last : pind - 1);
```

Reply:
I'm trying with both versions...

Reply:
If we keep the first proposed version (the one that passes the test as is), we need to update the documentation accordingly.

Reply:
It looks correct to me...
```
If no match found, function gives nearest element smaller than value.
```
Or did I miss something?

Reply:
In the code [snippet](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:32,endLineNumber:26,positionColumn:32,positionLineNumber:26,selectionStartColumn:32,selectionStartLineNumber:26,startColumn:32,startLineNumber:26),source:'%23include+%3Calgorithm%3E%0A%23include+%3Ciostream%3E%0A%23include+%3Cvector%3E%0A%0A///+Binary+search+in+an+array+defined+by+its+iterators.%0A///%0A///+The+values+in+the+iterators+range+are+supposed+to+be+sorted%0A///+prior+to+this+call.++If+match+is+found,+function+returns%0A///+position+of+element.++If+no+match+found,+function+gives+nearest%0A///+element+smaller+than+value.%0Atemplate+%3Ctypename+Iterator,+typename+Element%3E%0AIterator+BinarySearch(Iterator+first,+Iterator+last,+Element+value)%0A%7B%0A+++Iterator+pind%3B%0A+++pind+%3D+std::lower_bound(first,+last,+value)%3B%0A+++if+(+(pind+!!%3D+last)+%26%26+(*pind+%3D%3D+value)+)%0A++++++return+pind%3B%0A+++else%0A++++++return+((pind+%3D%3D+first)+%3F+pind+:+pind+-+1)%3B%0A%7D%0A%0A//+i.e.+if+the+smallest/first+element+is+larger+than+the+value,+%0A//+it+still+returns+!'first!'.%0Aint+main()%0A%7B%0A++++std::vector%3Cfloat%3E+in%7B+10.0,+20.0+%7D%3B%0A++++auto+iter+%3D+BinarySearch(+in.begin(),+in.end(),+1.0)%3B%0A++++if+(iter+%3D%3D+in.end())%0A++++++std::cout+%3C%3C+%22Returned+the+end+marker%5Cn%22%3B%0A++++else+%0A++++++std::cout+%3C%3C+%22Returned+an+iterator+pointing+to+the+value:%22%0A+++++++++%3C%3C+*iter+%3C%3C+!'%5Cn!'%3B%0A++++return+0%3B%0A%7D%0A%0A%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:51.145194274028626,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:g132,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B20+-Wall+-Wextra+-Wpedantic',overrides:!(),runtimeTools:!(),source:1,stdinPanelShown:'1',tree:0,wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+gcc+13.2+(C%2B%2B,+Editor+%231)',t:'0')),header:(),k:48.85480572597136,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) 
it shows that with the vector:
```
{ 10.0, 20.0 }
```
searching from `1.0`
returns `10.0` which is not "the nearest element smaller than value."

Reply:
OK, fine, I'll change the doc